### DISCUSSION TITLE
If Mir doesn't produce regular binaries than how can we use libraries written in other languages?

### DISCUSSION THREAD

**Discussion by rempas** [vnmakarov/mir#210] [Q&A] [answered]
So Mir produces a ".bmir" files that is a memory binary (if I remember how you said it in the previous post). So how can I combine it with an object file or a library (".so", ".a")?

**Comment by vnmakarov** [comment, marked as answer]
bmir is a binary mir representation.  It is basically another form of MIR file representation beside MIR textual file representation.  Binary form is more compact and much more faster to read than the textual form.

To get machine code you still need to generate it from MIR independently how you get it, e.g from reading binary or textual mir files or creating MIR in memory by MIR API (using funcs MIR_new_module, MIR_new_func, MIR_new_insn, etc).

To call external functions (e.g. from .so or .a library or code which uses libmir.a) from MIR code, your need to provide `import_resolver` function to MIR_link function call.   When MIR_link links MIR code and see external (imported) functions/data reference and cannot find such functions/data in already loaded MIR code, it calls `import_resolver` to get the address of the external function/data.
  
If you use external functions implemented on a language different from C, you also need to transform C call ABI (describing how args are passed) used by MIR to call ABI used by the implementation language.

**Reply by rempas** [reply]
Oh, I see! How can someone create a binary (that contains machine code) from MIR? Is this what the program "m2b" does? I was never able to make this program work. The same with "b2m".
